fairlearn/grid_search.py

```diff
@@ -0,0 +1,109 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+"""
+This module implements a 1d grid search for binary classification with a single
+binary protected attribute.
```

Comment:
Should we add a comment on how this won't scale beyond one or two dimensions?

Reply:
Done

Reply:
I'm not 100% convinced this is right. If I remember the paper correctly then even using a different fairness metric such as EO would mean that we get multiple dimensions, thus making it 2d (despite a single protected attribute). So if this is specific to DP then we should probably say so.
@MiroDudik might be able to comment.

Reply:
I mention in the actual 'work' method below that this is for demographic parity. That was based on arguing with myself that there should be more dimensions _somewhere_ if we were doing EO (not the most sophisticated argument, I know :-) ). Agreed that @MiroDudik  is the authority here :-)

Reply:
Yes. The current implementation is for: binary classification + demographic parity with a binary protected attribute. I think we should signal that in the name of the class below. How about something like "DemographicParityGridSearch", or even "GridSearchDemographicParity"... since I imagine we'll also have "GridSearchBoundedGroupLoss".

Reply:
Will we also be wanting a multiclass grid search? How about the following names:
```python
class BinaryAttributeGridSearchDemographicParity:
    def binary_classification(…):

    def multiclass_classification(…):

class BinaryAttributeGridSearchBoundedGroupLoss:
    def binary_classification(…):

    def multiclass_classification(…):

class BinaryAttributeGridSearchRegression:
    def regression(…):
```
Or perhaps the 'BinaryAttributeGridSearch' "stem" should be hoisted into the filename.

Reply:
Multiclass would also mean multiple dimensions, right?

Reply:
I don't think so.... @MiroDudik ? I had the impression that the dimensionality was set by the protected attribute, not the number of unique labels

Reply:
I think you're right (except for the EO case in which the labels would definitely factor in since we're conditioning on them, but we don't care about EO right now I think).

Reply:
I have a proposed class renaming, which I think will work for now